src/main/java/cloud/fogbow/ras/core/plugins/interoperability/cloudstack/compute/v4_9/CloudStackComputePlugin.java

```diff
@@ -108,164 +104,212 @@ public String requestInstance(ComputeOrder computeOrder, CloudStackUser cloudUse
+        throw new NoAvailableResourcesException(
+                Messages.Error.UNABLE_TO_COMPLETE_REQUEST_SERVICE_OFFERING_CLOUDSTACK);
+    }
+
+    @VisibleForTesting
+    List<GetAllServiceOfferingsResponse.ServiceOffering> filterServicesOfferingByRequirements(
+            @NotNull List<GetAllServiceOfferingsResponse.ServiceOffering> serviceOfferings,
+            @NotNull ComputeOrder computeOrder) {
 
-        for (GetAllServiceOfferingsResponse.ServiceOffering serviceOffering : serviceOfferings) {
-            if (serviceOffering.getCpuNumber() >= computeOrder.getvCPU() &&
-                    serviceOffering.getMemory() >= computeOrder.getMemory()) {
-                return serviceOffering;
-            }
+        List<GetAllServiceOfferingsResponse.ServiceOffering> serviceOfferingsFilted = serviceOfferings;
+        Map<String, String> requirements = computeOrder.getRequirements();
+        if (requirements == null || requirements.size() == 0) {
+            return serviceOfferings;
+        }
+
+        for (Map.Entry<String, String> tag : requirements.entrySet()) {
+            String tagFromRequirements = tag.getKey() + FOGBOW_TAG_SEPARATOR + tag.getValue();
+            serviceOfferingsFilted = serviceOfferingsFilted.stream().filter(serviceOffering -> {
+                String tagsServiceOffering = serviceOffering.getTags();
+                boolean isMatchingWithRequirements = tagsServiceOffering != null &&
+                        !tagsServiceOffering.isEmpty() &&
+                        tagsServiceOffering.contains(tagFromRequirements);
+                return isMatchingWithRequirements;
+            }).collect(Collectors.toList());
         }
 
-        return null;
+        return serviceOfferingsFilted;
     }
 
-    private GetAllServiceOfferingsResponse getServiceOfferings(CloudStackUser cloudUser) throws FogbowException {
-        GetAllServiceOfferingsRequest request = new GetAllServiceOfferingsRequest.Builder().build(this.cloudStackUrl);
-        CloudStackUrlUtil.sign(request.getUriBuilder(), cloudUser.getToken());
+    @NotNull
+    @VisibleForTesting
+    GetAllServiceOfferingsResponse getServiceOfferings(@NotNull CloudStackUser cloudStackUser)
+            throws FogbowException {
+
+        GetAllServiceOfferingsRequest request = new GetAllServiceOfferingsRequest.Builder()
+                .build(this.cloudStackUrl);
+        URIBuilder uriRequest = request.getUriBuilder();
+        String token = cloudStackUser.getToken();
+        CloudStackUrlUtil.sign(uriRequest, token);
 
-        String jsonResponse = null;
         try {
-            jsonResponse = this.client.doGetRequest(request.getUriBuilder().toString(), cloudUser);
+            String jsonResponse = doGet(uriRequest.toString(), cloudStackUser);
+            return GetAllServiceOfferingsResponse.fromJson(jsonResponse);
         } catch (HttpResponseException e) {
             CloudStackHttpToFogbowExceptionMapper.map(e);
+            return null;
```

Comment:
this line will never be executed once the line above breaks the normal workflow by throwing an exception

Reply:
check the same issue in the other methods please